### ISSUE TITLE
Writing augmented features to disk/option to perform data augmentation on cpu but everything else on gpu

### ISSUE THREAD

**Issue by cvqluu** [speechbrain/speechbrain#582]
Hi, first of all great project, everything that I've looked into seems quite readable.

I'm trying to play around with your VoxCeleb ECAPA TDNN recipe and I'm running out of GPU memory with the configured settings (which can obviously be solved by reducing the batch size).

However, if I lift the model architecture code and train it in on my own precomputed features (like from Kaldi) I can use the 'correct' batch size for the recipe with everything fitting on GPU(s).

My question is if there will be a plan to add the feature of writing some extracted feats to disk to save memory or alternatiely allowing just the augment_pipeline bit to be done on cpu.

Thanks,
Chau

**Comment by popcornell** [comment]
Hi, 

> My question is if there will be a plan to add the feature of writing some extracted feats to disk to save memory or alternatiely allowing just the augment_pipeline bit to be done on cpu.
the second option should could be already done using Dynamic Item Pipelines. 
See the tutorial on dataIO: https://colab.research.google.com/drive/1AiVJZhZKwEI4nFGANKXEe-ffZFfvXKwH
You could write a pipeline to do feature extraction as: 

@speechbrain.utils.data_pipeline.takes("file_path")
@speechbrain.utils.data_pipeline.provides("fbanks")
def fbank_pipeline(file_path):
      sig = speechbrain.dataio.dataio.read_audio(file_path)
      fbank = your_filterbank_function(sig) 
      return fbank 
      
   you could use speechbrain features or torchaudio or your own. This is done on CPU and is parallelized across workers. 
   

> My question is if there will be a plan to add the feature of writing some extracted feats to disk to save memory or alternatiely allowing just the augment_pipeline bit to be done on cpu.
Definitely we should have this for large datasets and network storage, I think this is related to issue https://github.com/speechbrain/speechbrain/issues/580.

**Comment by mravanelli** [comment]
Hi,
as for the OOM issue, the simplest solution is to reduce the batch size.
The parameters are set with our 32 GB GPU, buy you can easily reduce the
batch size.
As you can see, that recipe heavily relies on on-the-fly data augmentation
which really improves the performance. I thus strongly suggest feeding the
wav signals directly rather than pre-computed features. The latter
operation is possible (see Samuele's message), but there are good reason
for plugging the waveforms directly:


   1. Feature computation is super fast (especially on GPUs) and not memory
   demanding.
   2. You can easily combine on-the-fly feature computation with on-the-fly
   data augmentation
   3. The pipeline is cleaner and does not require annoying preliminary
   steps such as dumping features on disk.
   4. We can backpropagate the gradient through the features and make them
   learnable
   5. Conceptually, we just see feature extraction as an early layer of a
   deep architecture

**Comment by cvqluu** [comment]
Thanks for the quick answers! That makes a lot of sense, thanks.
